### PR TITLE
[Feature] Add gated Lance scan planning and full-dataset ranges

### LINKED ISSUE #74665: [Feature] Native Lance connector for StarRocks
- [x] https://github.com/StarRocks/starrocks/pull/74659
## Is your feature request related to a problem? Please describe.

StarRocks currently supports external table connectors for Iceberg, Delta Lake, Hudi, Paimon, and other formats. However, there is no support for [Lance](https://github.com/lancedb/lance), a modern columnar data format optimized for ML workloads — embedding vectors, multi-modal data, and large-scale training datasets.

Users who store datasets in Lance format must currently ETL data into a supported format before querying with StarRocks, adding latency and operational complexity.

## Describe the solution you'd like

A native Lance connector that allows StarRocks to directly query Lance datasets on local and cloud filesystems (S3, ADLS Gen2, GCS), enabling:

- Ad-hoc exploration of ML training datasets (embeddings, document corpora)
- Quality analysis by joining Lance datasets with metadata in Iceberg/Hive tables
- Interactive analytics replacing Spark read paths on Lance tables

## Implementation Plan

The implementation is split into independently reviewable PRs:

| PR | Scope | Status |
|----|-------|--------|
| #74659 | Thrift definitions for Lance scan range fields | Open |
| #74660 | FE optimizer operators and rules | Open |
| #74661 | FE scan node and plan fragment builder | Open |
| #74662 | Java-extensions Lance JNI reader | Open |
| #74663 | BE optional pluggable connector module (follows benchmark pattern) | Open |
| #74664 | Build flag `--without-connector-lance` | Open |

**Already merged (Phase 1):** FE catalog & metadata discovery (`LanceConnector`, `LanceMetadata`, `LanceApiConverter`, `LanceTable`)

## Architecture

- **FE**: `LANCE` registered as `ConnectorType` and `TableType`. Query planning reuses `THdfsScanRange` with Lance-specific fields. CBO statistics integration.
- **BE**: Optional pluggable module following the benchmark connector pattern (`--without-connector-lance` compile-time opt-out). JNI scanner delegates to `lance-core` Java SDK.
- **java-extensions**: `LanceSplitScanner` using Arrow vectors via `lance-core:7.0.0`

## Future Work

- Predicate pushdown expression visitor
- Cloud storage credential passthrough (Azure Workload Identity, S3 IAM, GCS)
- HMS-backed catalog discovery (replacing property-based bootstrap)
- Native Rust FFI scanner (zero-copy via Arrow C Data Interface)

## Are you willing to submit PR?

Yes — PRs are already submitted (see table above).

### PR DESCRIPTION
**Pull request by beinan**
## Why I'm doing:

Lance needs an executable FE fragment translator without exposing queries before the BE and packaged reader are available.

## What I'm doing:

Add the Lance scan node, table descriptor serialization, and fragment translator. SQL remains disabled until #74663 supplies the BE execution path.

Create one explicit whole-dataset range with no fragment-zero payload. Repeated range setup is idempotent; missing live compute nodes fail instead of fabricating a backend or returning an empty scan. Preserve the predicates serialized by `PlanNode`, rather than replacing them with the unpopulated Hive partition-pruning list.

The native multi-fragment reader regression is added in #74662; SQL plan tests are enabled with execution in #74663.

Series order: #74660 → #74661 → #74662 → #74663 → #74664. Catalog and Thrift prerequisites (#74420 and #74659) are already merged. These branches are cumulative against `main`; each preceding layer is required.

### Validation

All 20 focused FE optimizer, metadata, scan-node, and physical-plan tests pass at this PR head; Checkstyle also passes. Tests exercise projection/predicate/LIMIT translation, planner errors when workers are unavailable, shared-nothing and shared-data worker selection, repeated range setup, EXPLAIN output, and table/scan Thrift serialization.

A fresh local JaCoCo run covers 185 of the same 194 added executable FE lines counted by CI (95.36%, up from the previous CI result of 46.39%). All added lines in LanceScanNode and the Lance fragment translator are covered. Physical-plan tests invoke the translator directly, preserving the SQL gate in this preparatory layer.

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
  - [ ] This pr needs auto generate documentation
- [ ] This is a backport pr

Documentation for user-visible execution is included in #74663; this series targets `main` only.

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 4.1
  - [ ] 4.0
  - [ ] 3.5

## Related Issue

Fixes #74665

This PR implements one layer of the linked feature. The complete implementation requires all five PRs in the series listed above.

### CODE REVIEW

**Comment by CelerData-Reviewer** [comment]
@codex review

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java [hunk 1]
```diff
@@ -192,6 +193,7 @@
 import com.starrocks.sql.optimizer.operator.physical.PhysicalJDBCScanOperator;
 import com.starrocks.sql.optimizer.operator.physical.PhysicalJoinOperator;
 import com.starrocks.sql.optimizer.operator.physical.PhysicalKuduScanOperator;
+import com.starrocks.sql.optimizer.operator.physical.PhysicalLanceScanOperator;
```
[the remark this answers is not in the document]
Comment (beinan) [reply fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:196]: False positive: `PhysicalLanceScanOperator` is added in PR #74660 which is a prerequisite for this PR. See #74665.

File: fe/fe-core/src/main/java/com/starrocks/planner/LanceScanNode.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        if (nodeIds.isEmpty()) {
+            return;
+        }
+
+        TScanRangeLocations scanRangeLocations = new TScanRangeLocations();
+
+        THdfsScanRange hdfsScanRange = new THdfsScanRange();
+        hdfsScanRange.setUse_lance_jni_reader(true);
+        hdfsScanRange.setLance_dataset_uri(lanceTable.getUri());
+        hdfsScanRange.setLance_split_info("{\"fragment_ids\": [0]}");
```
[the remark this answers is not in the document]
Comment (beinan) [reply fe/fe-core/src/main/java/com/starrocks/planner/LanceScanNode.java]: False positive: Lance Thrift fields are added in PR #74659 which is a prerequisite for this PR. See #74665.

**Comment by beinan** [comment]
@github-actions run-checks

**Comment by beinan** [comment]
@dirtysalt @satanson Conflicts have been resolved and the Lance PR stack has been rebased onto the latest main. This PR follows #74660. Could you please take another look when available? CI is rerunning on the updated commits.

**Comment by beinan** [comment]
@dirtysalt @satanson The dependency-chain review comments have been clarified and all review threads are now resolved. This PR is currently mergeable. Please take another look when available.

**Comment by beinan** [comment]
Rebased onto the latest main (conflict-free). CI was already green on this PR; re-running on the updated commits to keep the stack consistent.

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by beinan** [comment]
@dirtysalt @satanson — second PR in the Lance connector stack (blocked on #74660, which is green and awaiting review).

Status: **CI fully green**, no conflicts, rebased onto latest main. The two Codex P1 comments were dependency-chain false positives — `PhysicalLanceScanOperator` comes from #74660 and the Lance thrift fields from the already-merged #74659 — and are resolved.

Scope: the scan node (`LanceScanNode`) and the `PlanFragmentBuilder` translator hookup.

Merge order: #74660 → **#74661** → #74662 → #74663 → #74664. No rush ahead of #74660, but an early look would help keep the stack moving.
